qiime2/core/validate.py

```diff
@@ -0,0 +1,105 @@
+        """
+        Used by Plugin to add a `ValidatorRecord` for a new validator to a
+        plugin. See plugin/plugin for definition of `ValidatorRecord`. Usually
+        called through the `register_validtor` decorator.
+        """
+        self._validators.append(validator_record)
+        self._is_sorted = False
+
+    def add_validation_object(self, *others):
+        """
+        Used to combine `ValidationObject`s for the same `concrete_type` from
+        different plugins(or just different objects. This is done
+                non-heirarchically by `PluginManager` by creating a new, blank
+                object for each `concrete_type` that it encounters, then adds
+                the `ValidationObject`s from each plugin.
+        """
+        for other in others:
+            self._validators += other._validators
+        self._is_sorted = False
+
+    @property
+    def validators(self) -> list:
+        """
+        'Public' facing way to access validators, this insures that a sorted
+        list is returned so the actual validation can be run.
+        """
+        if not self._is_sorted:
+            self._sort_validators()
+
+        return self._validators
+
+    def _sort_validators(self):
+        """
+        A partial order sort that will return a sorted list of validators. The
+        runtime for this sort is \theta(n^2). This is not a concern, as the
+        number of validators present for any particular type is expected to
+        remain trivially low. The validators are sorted from general to
+        specific.
+        """
+        self._validators = sorted_poset(
+            iterable=self._validators,
+            key=lambda record: record.context,
+            reverse=True)
+
+        self._is_sorted = True
+
+    def __call__(self, data, level):
+        """
+        Runs all validators stored in object on the provided data. Use of
+        `level` is required but the behaviour is defined in the individual
+        validators.
+        """
+        from_mt = ModelType.from_view_type(type(data))
+
+        for record in self.validators:
+            to_mt = ModelType.from_view_type(record.view)
+            transformation = from_mt.make_transformation(to_mt)
+            data = transformation(data)
+            try:
+                record.validator(data=data, level=level)
+            except ValidationError:
+                raise
+            except Exception as e:
+                raise ImplementationError("An unexpected error occured when %r"
+                                          " from %r attempted to validate %r"
+                                          % (record.validator.__name__,
+                                             record.plugin,
+                                             data)) from e
+
+    def assert_transformation_available(self, dir_fmt):
+        """
+        Called by `qiime2.sdk.PluginManager._consistency_check` to ensure that
```

Comment:
This is probably better as a comment, as it doesn't exactly explain how to use this public method (but the context is probably handy for the maintainer).